ch-image: add arch support (#1024)
hpc/charliecloud: doc/ch-image_desc.rst

   $ ch-image [...] list
   $ ch-image [...] pull [...] IMAGE_REF [IMAGE_DIR]
   $ ch-image [...] push [--image DIR] IMAGE_REF [DEST_REF]

>    $ ch-image [...] rename IMAGE_REF NEW_IMAGE_REF

Reviewer: Shouldn't `rename` be a separate PR?

  :code:`-v`, :code:`--verbose`
    Print extra chatter; can be repeated.

> 

Reviewer: Good catch.

Options:

>   :code:`--arch`
>     Specify architecture :code:`--arch=ARCH[/VARIANT]` for image described by
>     :code:`IMAGE_REF`. If found pull it; otherwise error. Use
>     :code:`--list-arch IMAGE_REF` for list of valid platform architectures.

Reviewer: I think it would be good, if we can, to avoid separating the architectures into variants, and keep it at a single level, i.e., some architectures happen to include a slash.

```suggestion
  :code:`--arch`
    Specify architecture :code:`--arch=ARCH` for image described by
    :code:`IMAGE_REF`. If image exists but not as architecture
    :code:`ARCH`, fail with error.
```

>   :code:`--list-arch`
>     Print a list of valid :code:`--arch` targets for image tagged
>     :code:`IMAGE_REF`.

Reviewer: ```suggestion
  :code:`--list-arch`
    Print list of available architectures for image
    :code:`IMAGE_REF`.
```

    Parse :code:`IMAGE_REF`, print a parse report, and exit successfully
    without talking to the internet or touching the storage directory.

This script tries to target the appropriate architecture when pulling. For
example, when pulling the image tagged :code:`centos:7`, first check for the

> existence of a fat manifest. If found, check for the architecture matching the

Reviewer: Can we avoid mentioning manifests and fat manifests?

host. If a match is found, pull it; otherwise proceed as if the fat manifest
does not exist. If a fat manifest does not exist, pull the image without trying

> to guess the appropriate architecture.

Reviewer: E.g. I don't know what this means. What architecture do I get? This should explain what the results are, not what's happening under the hood.

   RUN cp /opt/lib/*.so /usr/local/lib  # possible workaround
   RUN ldconfig

:code:`list`
------------

List images in storage.

::

   $ ch-image list
   00_tiny
   alpine:3.9
   alpine:latest
   centos:7
   debian:buster

:code:`list-arch`
-----------------

Print list of available architecturess for the Debian Buster image.::

   $ ch-image list-arch debian:buster
   listing architectures: debian:buster
   manifest list: using existing
   amd64
   arm/v5
   arm/v7
   arm64/v8
   386
   mips64le
   ppc64le
   s390x

:code:`pull`
------------

Download the Debian Buster image and place it in the storage directory::

   $ ch-image pull debian:buster
   pulling image:   debian:buster

>    architecture:    amd64

Reviewer: Looking at the current code, I think it qualifies as hard, since we'd have to carry around the source to a lot of placed. IMO we should omit it.

   bin   dev  home  lib64  mnt  proc  run   srv  tmp  var
   boot  etc  lib   media  opt  root  sbin  sys  usr

Print list of targetable architecturess for the Debian Buster image.::

   $ ch-image pull --list-arch debian:buster
   fat manifest: using existing
   available platforms:
   amd64
   arm/v5
   arm/v7
   arm64/v8
   386
   mips64le
   ppc64le
   s390x

> Specify the aarch64 version of Debian Stretch using :code:`--arch`.::

Reviewer: ```suggestion
Specify the aarch64 version of Debian Stretch using :code:`--arch`::
```

Author: The other examples in this file have a period. Should I remove them as well?

   $ ch-grow pull --arch=arm64/v8 debian:buster
   pulling image:   debian:buster
   fat manifest: using existing

>    architecture: arm64v8 (aarch64)

Reviewer: What does it mean to have two architectures listed?

Author: Ah, good question. This shows us both what Docker calls the architecture and what something like `uname -m` reports. Since they can differ, e.g., above and `amd64 == x86_64`, I thought it may be helpful if both were included in the output.

Reviewer: Aha. I think usability wise, we ought to stay in Docker’s vocabulary for these. Later when we’re guessing architectures we can show the translation, but here I think showing two different archs is confusing, it means we have to make the map available more broadly in the code, and going both directions will add complexity if it turns out it’s not 1-to-1.

   manifest: downloading
   layer 1/1: 04aacb1: downloading
   layer 1/1: 04aacb1: listing
   validating tarball members
   resolving whiteouts
   flattening image
   layer 1/1: 04aacb1: extracting
   done

> Specify the aarch32 version of Debian Stretch using long name.::

Reviewer: Would anyone want to do this?

Author: Do what exactly? Pull a aarch32 image or pull something different than the host they're on?

   cleaning up
   done

> :code:`list`

Reviewer: Put in alphabetical order.

-----------

List images in storage.

::

>   $ ch-image list
>   IMAGE REFERENCE            NAME      TAG        ARCH
>   arm64/v8/debian:stretch    debian    stretch    arm64v8 (aarch64)
>   debian:stretch             debian    stretch    amd64 (x86_64)
>   centos:8                   centos    8          ppcle64
>   centos:7                   centos    7          amd64 (x86_64)

Reviewer: This is going to be hairy to compute the column widths, and because we don't currently distinguish between name and tag. Possibly:

```suggestion
  $ ch-image list
  arm64/v8/debian:stretch (aarch64)
  debian:stretch (x86_64)
  centos:8 (ppcle64)
  centos:7 (x86_64)
```

Also, why do we have one `debian:stretch` with an architecture path and one without?

Author: The image reference is unique. In these examples the image with the arch path was pulled via long-name, e.g., `ch-image pull arm64/v8/debian:stretch` thus exists in storage as `arm64%v8d%ebian:stretch`. 

This also illustrates a key assumption: that architecture guessing is only for short-names, e.g., `debian:stretch`. If you specify a long-name you know what you're doing and what you're asking for. For example, it doesn't make much sense for me to do `ch-image pull ppc64le/centos:8` and get a `x86_64` version.

Reviewer: Aha. I don't follow the distinction between short and long names. Seems too magic to me.

Reviewer: E.g. what about `foo/bar/debian:stretch`? It's a valid image name but how do we know `foo/bar` isn't an architecture?
